### PR TITLE
Refactor SequentialReader to use pimpls

### PR DESCRIPTION
**Pull request by piraka9011**
## Changes
* Refactored implementation of `SequentialReader` into `src/rosbag2/impl/sequential_reader_impl.hpp`
* Exposed `SequentialReader` interface through `sequential_reader.hpp` only.

## Issue
 https://github.com/ros-security/aws-roadmap/issues/97

### CODE REVIEW

File: rosbag2/include/rosbag2/sequential_reader.hpp [hunk 1]
```diff
@@ -100,10 +101,8 @@ class ROSBAG2_PUBLIC SequentialReader
   virtual std::vector<TopicMetadata> get_all_topics_and_types();
 
 private:
-  std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory_;
-  std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory_;
-  std::shared_ptr<rosbag2_storage::storage_interfaces::ReadOnlyInterface> storage_;
-  std::unique_ptr<Converter> converter_;
+  class ReaderImpl;
+  std::unique_ptr<ReaderImpl> reader_pImpl_;
```
Comment (thomas-moulard) [inline rosbag2/include/rosbag2/sequential_reader.hpp]: The names of variables (including function parameters) and data members are all lowercase, with underscores between words. Data members of classes (but not structs) additionally have trailing underscores. For instance: a_local_variable, a_struct_data_member, a_class_data_member_.

https://google.github.io/styleguide/cppguide.html#Variable_Names

File: rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace rosbag2
+{
+
+class SequentialReader::ReaderImpl
```
Comment (thomas-moulard) [inline rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp]: (minor) it'd be preferable to have a cpp file for this one too

File: rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+#include "rosbag2/sequential_reader.hpp"
+
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <utility>
+#include <vector>
+
+namespace rosbag2
+{
```
Comment (thomas-moulard) [inline rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp:25]: it's preferable to put "private" local classes in an anonymous namespace

https://www.ibm.com/support/knowledgecenter/en/ssw_ibm_i_72/rzarg/unnamed_namespaces.htm

File: rosbag2/include/rosbag2/sequential_reader.hpp [hunk 4]
```diff
@@ -100,10 +101,8 @@ class ROSBAG2_PUBLIC SequentialReader
   virtual std::vector<TopicMetadata> get_all_topics_and_types();
 
 private:
-  std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory_;
-  std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory_;
-  std::shared_ptr<rosbag2_storage::storage_interfaces::ReadOnlyInterface> storage_;
-  std::unique_ptr<Converter> converter_;
+  class ReaderImpl;
```
Comment (thomas-moulard) [inline rosbag2/include/rosbag2/sequential_reader.hpp]: I'd put that out of this class so that it can go into an anonymous namespace

**Review by thomas-moulard — APPROVED** [review]
This LGTM, a couple nits.

File: rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp (hunk 2 above)
Comment (piraka9011) [reply rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp]: Sure, keep under `src/rosbag2/impl` or just `src/rosbag2`?

Comment (thomas-moulard) [reply rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp]: everything in src is impl :)
I'd just use src/rosbag2

File: rosbag2/include/rosbag2/sequential_reader.hpp [hunk 5]
```diff
@@ -102,7 +102,7 @@ class ROSBAG2_PUBLIC SequentialReader
 
 private:
   class ReaderImpl;
-  std::unique_ptr<ReaderImpl> reader_pImpl_;
+  std::shared_ptr<ReaderImpl> reader_pimpl_;
```
Comment (thomas-moulard) [inline rosbag2/include/rosbag2/sequential_reader.hpp]: This must be unique_ptr, we should not leak this object.

**Review by thomas-moulard — CHANGES_REQUESTED** [review]

File: rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+   * format of the storage factory.
+   *
+   * \param converter_serialization_format
+   * \param storage_serialization_format
+   */
+  virtual void check_converter_serialization_format(
+    const std::string & converter_serialization_format,
+    const std::string & storage_serialization_format);
+
+  std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory_;
```
Comment (thomas-moulard) [inline rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp]:
```suggestion
  std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory_{};
```

File: rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+  std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory_;
```
Comment (thomas-moulard) [inline rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp]:
```suggestion
  std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory_{};
```

File: rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+  std::shared_ptr<rosbag2_storage::storage_interfaces::ReadOnlyInterface> storage_ {nullptr};
```
Comment (thomas-moulard) [inline rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp]:
```suggestion
  std::shared_ptr<rosbag2_storage::storage_interfaces::ReadOnlyInterface> storage_ {};
```

File: rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+  std::unique_ptr<Converter> converter_ {nullptr};
```
Comment (thomas-moulard) [inline rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp]:
```suggestion
  std::unique_ptr<Converter> converter_ {};
```

File: rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp (hunk 3 above)
Comment (thomas-moulard) [reply rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp:25]: still not in an anonymous ns?

File: rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+
+  ~ReaderImpl();
+  void reset();
+  void open(
+    const StorageOptions & storage_options, const ConverterOptions & converter_options);
+  bool has_next();
+
+  std::shared_ptr<SerializedBagMessage> read_next();
+
+  std::vector<TopicMetadata> get_all_topics_and_types();
```
Comment (thomas-moulard) [inline rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp]: Not for this PR, bur we should have ticket to fix those methods const-ness.

File: rosbag2/src/rosbag2/impl/sequential_reader_impl.cpp [hunk 11]
```diff
@@ -0,0 +1,114 @@
...
+
+#include "sequential_reader_impl.hpp"
+
+namespace rosbag2
+{
+
+ReaderImpl::ReaderImpl(
+  std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory,
+  std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory)
+: storage_factory_(std::move(storage_factory)),
```
Comment (thomas-moulard) [inline rosbag2/src/rosbag2/impl/sequential_reader_impl.cpp:29]: Are you sure this works? I believe you'll need a private default ctor to be called here before you overwrite the attributes. I think that as soon as you have a non-default ctor, the inline attribut init it not called anymore.

File: rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp [hunk 12]
```diff
@@ -0,0 +1,74 @@
...
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "rosbag2/sequential_reader.hpp"
+
+namespace rosbag2
+{
+
+class ReaderImpl
```
Comment (thomas-moulard) [inline rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp]: Let's make this LOCAL -> https://github.com/aws-ros-dev/rosbag2/blob/master/rosbag2/include/rosbag2/visibility_control.hpp#L40

**Review by thomas-moulard — COMMENTED** [review]
This LGTM again with the unique_ptr, a few minor nits.

File: rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp (hunk 6 above)
Comment (piraka9011) [reply rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp]: Do we still need this if there is a default in the constructor?

File: rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp (hunk 3 above)
Comment (thomas-moulard) [reply rosbag2/src/rosbag2/impl/sequential_reader_impl.hpp:25]: as we discussed offline, anonymous ns is not possible now that facade and impl are in two different compilation units.

**Review by thomas-moulard — APPROVED** [review]
LGTM w/ local added

**Comment by piraka9011** [comment]
Opened PR against [master](https://github.com/ros2/rosbag2/pull/191)

**Pull request closed without merging**